diff --git a/napi-inl.h b/napi-inl.h
index 568a6ba..c0a8389 100644
--- a/napi-inl.h
+++ b/napi-inl.h
@@ -13,11 +13,13 @@
 #include "napi.h"

 #include <algorithm>
+#include <atomic>
 #include <cstdarg>
 #include <cstring>
 #if NAPI_HAS_THREADS
 #include <mutex>
 #endif  // NAPI_HAS_THREADS
+#include <string>
 #include <type_traits>
 #include <utility>

@@ -59,13 +61,16 @@ inline napi_status AttachData(napi_env env,
                               FreeType* data,
                               void* hint = nullptr) {
   napi_status status;
-#if (NAPI_VERSION < 5)
+#if (NAPI_VERSION < 5) || defined(OHOS_PLATFORM)
+  static std::atomic_uint64_t globalPropCounter{0};
   napi_value symbol, external;
   status = napi_create_symbol(env, nullptr, &symbol);
   if (status == napi_ok) {
     status = napi_create_external(env, data, finalizer, hint, &external);
     if (status == napi_ok) {
-      napi_property_descriptor desc = {nullptr,
+      const auto propSeqNo = ++globalPropCounter;
+      const auto propName = "__Napi::AttachData_" + std::to_string(propSeqNo);
+      napi_property_descriptor desc = {propName.c_str(),
                                        symbol,
                                        nullptr,
                                        nullptr,
@@ -1335,7 +1340,7 @@ inline std::string String::Utf8Value() const {
   value.reserve(length + 1);
   value.resize(length);
   status = napi_get_value_string_utf8(
-      _env, _value, &value[0], value.capacity(), nullptr);
+      _env, _value, &value[0], value.capacity(), &length);
   NAPI_THROW_IF_FAILED(_env, status, "");
   return value;
 }
@@ -1350,7 +1355,7 @@ inline std::u16string String::Utf16Value() const {
   value.reserve(length + 1);
   value.resize(length);
   status = napi_get_value_string_utf16(
-      _env, _value, &value[0], value.capacity(), nullptr);
+      _env, _value, &value[0], value.capacity(), &length);
   NAPI_THROW_IF_FAILED(_env, status, NAPI_WIDE_TEXT(""));
   return value;
 }
@@ -2142,15 +2147,17 @@ inline SharedArrayBuffer SharedArrayBuffer::New(napi_env env,

 inline void* SharedArrayBuffer::Data() {
   void* data;
-  napi_status status = napi_get_arraybuffer_info(_env, _value, &data, nullptr);
+  size_t length;
+  napi_status status = napi_get_arraybuffer_info(_env, _value, &data, &length);
   NAPI_THROW_IF_FAILED(_env, status, nullptr);
   return data;
 }

 inline size_t SharedArrayBuffer::ByteLength() {
+  void* data;
   size_t length;
   napi_status status =
-      napi_get_arraybuffer_info(_env, _value, nullptr, &length);
+      napi_get_arraybuffer_info(_env, _value, &data, &length);
   NAPI_THROW_IF_FAILED(_env, status, 0);
   return length;
 }
@@ -2249,15 +2256,17 @@ inline ArrayBuffer::ArrayBuffer(napi_env env, napi_value value)

 inline void* ArrayBuffer::Data() {
   void* data;
-  napi_status status = napi_get_arraybuffer_info(_env, _value, &data, nullptr);
+  size_t length;
+  napi_status status = napi_get_arraybuffer_info(_env, _value, &data, &length);
   NAPI_THROW_IF_FAILED(_env, status, nullptr);
   return data;
 }

 inline size_t ArrayBuffer::ByteLength() {
+  void* data;
   size_t length;
   napi_status status =
-      napi_get_arraybuffer_info(_env, _value, nullptr, &length);
+      napi_get_arraybuffer_info(_env, _value, &data, &length);
   NAPI_THROW_IF_FAILED(_env, status, 0);
   return length;
 }
